### ISSUE TITLE
Keyframe is requested from peers too early when using REST interface.

### ISSUE THREAD

**Issue by vi** [jitsi/jitsi-videobridge#553]
According to [rest-colibri.md](../blob/master/doc/rest-colibri.md), one should send two PATCH requests to the video bridge's REST interface: the first gets bridge's credentials, the second sends peer's credentials. Actual ICE connection gets established after the second one.

In order to provide video streams to the newly connected peer, Videobridge requests keyframes from already connected peers (FIR). But happens on the first PATCH requests, when new peer is not yet fully joined and not on the second PATCH.

How do I properly join a peer and immediately receive incoming video, without waiting for eventual another keyframe?

**Comment by vi** [comment]
Looks like setting the `"last-n"` at the very end (using a third PATCH request) resolves the problem.

**Comment by jitsi-developers** [comment]
I actually thought this was a scenario the bridge was handling (by caching
some of the incoming data, I think?).  But Boris or George would know for
sure.

**Comment by bgrozev** [comment]
I don't know off the top of my head. I'll try to take a look and let you know in the following days.

**Comment by vi** [comment]
If LastN is expected to be set by REST user at the very end of the joining process then the documentation (rest-colibri.md) should be fixed to include the third PATCH.

**Comment by bgrozev** [comment]
> If LastN is expected to be set by REST user at the very end of the joining process then the documentation (rest-colibri.md) should be fixed to include the third PATCH.

No, I don't think that's the expectation. It should default to last-n=-1 (send all streams), which should result in correct behavior for keyframe requests.

**Comment by gpolitis** [comment]
> Looks like setting the "last-n" at the very end (using a third PATCH request) resolves the problem.

@bbaldino is right, this is a workaround for some other underlying issue. Normally the bridge should take care of this automatically.

**Comment by gpolitis** [comment]
> If LastN is expected to be set by REST user at the very end of the joining process then the documentation (rest-colibri.md) should be fixed to include the third PATCH.

That's not the expectation. The reason why this workaround solves the problem for you is because setting the last-n value causes an update to the bitrate allocator which results in a FIR.

There's a bunch of places in VideoChannel.java where we update the bitrate controller, for example when the pinned/selected endpoints change, when the streams are signaled, when the channel is initialized, when last-n is set, etc.

In your call flow the update isn't triggered, for some reason. I would expect it to be called when the channel is initialized or when you signal the streams.

**Comment by gpolitis** [comment]
@vi or anybody else who experiences this issue please provide a minimal reproducible scenario where this happens (including the full HTTP requests)? If it gets long feel free to go through support@jitsi.org.

**Comment by iwashi** [comment]
I don't have the minimal scenario right away but at least I'm experiencing the same issue. With REST API it takes a long time to see the video media stream flowing.

**Comment by marcusspangenberg** [comment]
I created PR #583 for a potential fix to this issue. A description of the change is in the PR.

**Comment by marcusspangenberg** [comment]
Closing the PR, since the issue still happens. Adding the description here, since it should give an indication to why this happens. The problem is still that for recvonly video channels, the video bridge fails to generate PLIs from time to time.

When using the rest api, no PLIs are sent by the video bridge to existing endpoints when a new channel is allocated. Client A allocates a recvonly audio channel and a sendrecv video channel. Client B allocates a recvonly audio channel and a recvonly video channel. When client B patches the bridge with its answer colibri, the video bridge should send a PLI to client A. This happens when BitrateController.update() is called during VideoChannel initialization. However this doesn't happen in the described scenario.

The call to BitrateController.update() in VideoChannel.initialize() does not generate a PLI, since the endpoint property of the dest VideoChannel is null. BitrateController.prioritize() returns null if the dest endpoint is unset. The other possible call to BitrateController.update() happens in VideoChannel.setRtpEncodingParameters(). However, since the recvonly channel has no sources, the "changed" boolean will always be false, and never call BitrateController.update().

It has bee suggested to add a dummy ssrc to recvonly channels to mitigate the issue. VideoChannel.setRtpEncodingParameters() will then call BitrateController.update(), but it still doesn't generate a PLI since the sourceEncodings[0].isActive() call in SimulcastController.setTargetIndex() returns false for the dummy stream.

**Comment by sarumjanuch** [comment]
I feel that we do have a very similar issue, but we are using colibri for signaling.

**Comment by gpolitis** [comment]
This is likely fixed in https://github.com/jitsi/jitsi-videobridge/pull/735
